# Fundamentals of Data Analysis Tasks

Phelim Barry

***

# Introduction

The purpose of this...

# Table of Contents

1. [Task1](#Task1)
2. [Task2](#Task2)
3. [Task3](#Task3)
4. [Task4](#Task4)
5. [Task5](#Task5)
6. [References](#References)



# Task1

>The Collatz conjecture is a famous unsolved problem in mathematics. The problem is to prove that if you start with any positive integer $x$ and repeatedly apply the function $f(x)$ below, you always get stuck in the repeating sequence 1, 4, 2, 1, 4, 2, . . .   
This task is to verify, using Python, that the conjecture is true for the first 10,000 positive integers.

If $x$ is even, $f(x) = x ÷ 2$   
If $x$ is odd, $f(x) = 3x + 1$

Define a function for $f(x)$ and apply the appropriate formula. Using the modulus operator we can determine if $x$ is even (i.e. no remainder when divided by 2) or else $x$ must be odd.

In [64]:
def f(x):
    if x % 2 == 0:          #x is even
        return x // 2
    else:                   #x is odd
        return (3 * x) + 1

Define a function to check the $x$ value. If after applying the formula to the number, we arrive at a value equal to 1 then we have proved for that number and move on to the next number. If we reach the number 10001 then we have proved for all numbers between 1 and 10000

In [65]:
def collatz(x):
    while x != 1:
        x = f(x)            #call the f(x) function

Run the function and print a confirmation message once complete.

In [66]:
# Define start and end range values
start_range = 1
stop_range = 10001

for num in range (start_range, stop_range):
    collatz(num)        #call the collatz function

print (f'We have cycled through the values from {start_range} to {stop_range-1} and each time ended with a value of 1 \nindicating the conjecture is true for the first {stop_range-1} positive integers')


We have cycled through the values from 1 to 10000 and each time ended with a value of 1 
indicating the conjecture is true for the first 10000 positive integers


# Task2


>The purpose of this task is to give an overview of the penguins data set explaining the types of variables it contains.

>We will also suggest the types of variables that should be used to model them in Python giving explanations for the rationale.

**Overview**

The Palmer Penguin Dataset contains measurements of three different species of penguins from three islands in the Palmer Archipelago in Antartica. The data were collected between 2007 and 2009 and made available by researcher Dr. Kristen Gorman and other researchers from the Palmer Research Station.

Similar to the Iris dataset it is often used to explore areas of data analysis such as correlation and regression.
 

The dataset is available to view in a number of locations where we could quickly look and see the column names such as from Michael Waskom's $^1$ repository on GitHub. From there we can see that the dataset has seven columns of data containing the following variables:

| Variable Name |
| --- |
| species |
| island |
| bill_length_mm |
| bill_depth_mm |
| flipper_length_mm |
| body_mass_g |
| sex |
 

The dataset is also available directly through seaborn so by importing it we can do some analysis on the data to better understand the amount of data and the different columns.

In [67]:
# Import seaborn and load the penguins data set into a new dataframe df

import seaborn as sns
import numpy as np

df=sns.load_dataset ('penguins')


In [68]:
# print the dataframe to view the data

df


species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
339  Gentoo     Biscoe             NaN            NaN                NaN   
340  Gentoo     Biscoe            46.8           14.3              215.0   
341  Gentoo     Biscoe            50.4           15.7              222.0   
342  Gentoo     Biscoe            45.2           14.8              212.0   
343  Gentoo     Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
0         3750.0    Male  
1         3800.0  Female  
2         3250.0  Female  
3            NaN     NaN  
4         3450.0  Female  
..           ...     ...  
339          NaN     NaN  
340       4850.0  Female  
341       5750.0    Male  
342       5200.0  Female  
343       5400.0    Male  

[344 rows x 7 columns]

From this we can see that species, island and sex contain categorical data while the remaining seem to contain numerical data.

We can then use the ```unique``` command to list the full set of values in each column containing categorical variables...

In [69]:
#x = df.species.unique()
print(f'The values in species column are {df.species.unique()}')
print(f'The values in island column are {df.island.unique()}')
print(f'The values in sex column are {df.sex.unique()}')


The values in species column are ['Adelie' 'Chinstrap' 'Gentoo']
The values in island column are ['Torgersen' 'Biscoe' 'Dream']
The values in sex column are ['Male' 'Female' nan]


Categorical variable types are used when the variable can only be one of a specific list of values or used to describe qualitative data. In the penguins dataset the three categorical variables have very specific values.   
Species has three values: Adelie, Chinstrap and Gentoo. Island also has three values: Biscoe, Dream and Torgersen. Sex has two values: male and female.   
All three of these variables can be described as nominal as they are being used to name something.   



Numeric variable types are also known as quantitative variables because they typically are used to measure something using a number value. The penguins dataset contains four numeric variables with three measured in millimeters and one measured in grams. These are continuous variables and can be further broken down and described as ratio variables because they cantain numbers that have measurable differences that can be determined such as the differences in lengths, weights etc. Given that all four of the variables are being used to measure something they would be considered to be using real numbers. The presence of decimal places also points to them being real numbers.

Based on the analysis above we can determine that the dataset contains the following variables and data types.

| Variable Name | Type | Description |
| --- | --- | --- |
| species | categorical | penguin species - Adelie, Chinstrap or Gentoo |
| island | categorical | island name - Biscoe, Dream or Torgersen |
| bill_length_mm | numeric/real | penguin bill length measured in mm |
| bill_depth_mm | numeric/real | penguin bill depth measured in mm |
| flipper_length_mm | numeric/real | penguin flipper length measured in mm |
| body_mass_g | numeric/real | penguin body mass measured in grams |
| sex | categorical | penguin sex - male or female |

**Suggested Variable Types in Python**

To model each of the variables in Python the following data types are suggested...


| Variable Name | Type | Python Variable Type |
| --- | --- | --- |
| species | categorical | string |
| island | categorical | string |
| sex | categorical | string |
| bill_length | numeric/real | float |
| bill_depth | numeric/real | float |
| flipper_length | numeric/real | float |
| body_mass | numeric/real | float |

Categorical variables typically only contain a limited number of possible values and would not be used to perform any mathematical operations. They can contain alpha characters, numerica values or even special characters but usually would not contain numbers only. Such variables are stored as type ```string``` or ```str``` in Python.   

Species, Island and sex can all be modeled as type ```str```. All three variables contain alpha characters and are a small set of fixed values. 

Numeric values in python usually are defined as type ```int``` or ```floating point``` (can also be complex). ```int``` variable types are whole numbers with no decimal places. They are ideal for performing addition or subtraction but if we are to perform any arithmetic involving division such as calculating means or average values, then our calculated value would turn into a ```float``` by default python logic. ```float``` or floating-point variables are used to store values with decimal points. Arithmetic on float values will always result in another float value and will display a value after the decimal point even if the value is $.0$.   

With this in mind it would be best to model bill_length and bill_depth as type ```float``` because both contain decimal values. 
flipper_length and body_mass could be stored as ```int``` variables as they do not contain decimals however if we needed to convert the values to centimeters or kilograms then that would introduce decimal places so it would be best to store them as ```float``` also.

In total, there are 344 rows of data in the dataset. However, unlike the iris dataset, the penguins dataset has some missing values which show up as 'NaN' when the data is imported. These can be found using the ```isnull().sum()``` command as follows:

In [70]:
df.isnull().sum()


species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
dtype: int64

When analysing the data these will need to be taken into consideration and dealt with appropriately. For example, 

# Task3

>For each of the variables in the penguins data set, suggest what probability distribution from the numpy random distributions list is the most appropriate to model the variable.

The dataset contains the following variables:
| Variable Name |
| --- |
| species |
| island |
| bill_length_mm |
| bill_depth_mm |
| flipper_length_mm |
| body_mass_g |
| sex |

### Species ###  
We can see that the species variable is categorical and can only contain one of three possible values - Gentoo, Adelie or Chinstrap. The data shows that the numbers are not equally distributed   

    Adelie - 152 (44%)   
    Gentoo - 68 (20%)   
    Chinstrap - 124 (36%)      

To model this data we would use a multinomial distribution and include weighting to closely replicate the original data accordingly. 


In [71]:
# model species variable using multinomial distribution
rng = np.random.default_rng()
species_prob = rng.multinomial(10000, [0.44, 0.20, 0.36])

print ('Count = 10000')
print (f'Random number of Adelie = {species_prob[0]}')
print (f'Random number of Gentoo = {species_prob[1]}')
print (f'Random number of Chinstrap = {species_prob[2]}')

Count = 10000
Random number of Adelie = 4403
Random number of Gentoo = 2047
Random number of Chinstrap = 3550


### Island ###  
Similar to the species variable, the island variable is also categorical and again can only contain one of three possible values - Biscoe, Dream or Torgersen. The data shows that the numbers are not equally distributed   

    Biscoe - 168 (49%)   
    Dream - 124 (36%)   
    Torgersen - 52 (15%)      

To model this data we would again use a multinomial distribution and include weighting to closely replicate the original data accordingly. 

In [72]:
# model island variable using multinomial distribution
rng = np.random.default_rng()
island_prob = rng.multinomial(10000, [0.49, 0.36, 0.15])

print ('Count = 10000')
print (f'Random number of Biscoe = {island_prob[0]}')
print (f'Random number of Dream = {island_prob[1]}')
print (f'Random number of Torgersen = {island_prob[2]}')


Count = 10000
Random number of Biscoe = 4905
Random number of Dream = 3645
Random number of Torgersen = 1450


### Sex ###  
We can see that the sex variable only contains two possible values - male or female - meaning the data is discrete, and each sample is independent of each other. We can also see that there seems an almost equal possibility of a penguin being either male or female.

    Female - 165 (49.55%)   
    Male - 168 (50.45%)       


With these details in mind we could model the data using a binomial distribution.

In [73]:
# model sex variable using binomial distributions
rng = np.random.default_rng()

n, p = 1, .5045  # n = number of trials (success=male, failure=female), p = probability of each trial (.5045 for male)
sex_prob = rng.binomial(n, p, 10000)

# Print results
male = np.sum(sex_prob == 1)
female = np.sum(sex_prob == 0)
print ('count = 10000')
print (f'Random number of males = {male}')
print (f'Random number of females = {female}')

count = 10000
Random number of males = 4983
Random number of females = 5017


### Body Mass

Body Mass is a continuous variable and quite often the weights of a population are normally distributed meaning we could potentially model it using the 'normal distribution'. However if we look at the data we can see the difference between the mean and 50th percentile is significant and the deviations between min/max and the mean indicate some amount of skewness in the data.

In [74]:
df.body_mass_g.describe()

count     342.000000
mean     4201.754386
std       801.954536
min      2700.000000
25%      3550.000000
50%      4050.000000
75%      4750.000000
max      6300.000000
Name: body_mass_g, dtype: float64

We know there are different species of penguin in the data set so next step would be to look at each species individually... 

In [75]:
gentoo_df = df.loc[df.loc[:,'species'] == 'Gentoo']
gentoo_df.body_mass_g.describe()

count     123.000000
mean     5076.016260
std       504.116237
min      3950.000000
25%      4700.000000
50%      5000.000000
75%      5500.000000
max      6300.000000
Name: body_mass_g, dtype: float64

In [76]:
chinstrap_df = df.loc[df.loc[:,'species'] == 'Chinstrap']
chinstrap_df.body_mass_g.describe()

count      68.000000
mean     3733.088235
std       384.335081
min      2700.000000
25%      3487.500000
50%      3700.000000
75%      3950.000000
max      4800.000000
Name: body_mass_g, dtype: float64

In [77]:
adelie_df = df.loc[df.loc[:,'species'] == 'Adelie']
adelie_df.body_mass_g.describe()

count     151.000000
mean     3700.662252
std       458.566126
min      2850.000000
25%      3350.000000
50%      3700.000000
75%      4000.000000
max      4775.000000
Name: body_mass_g, dtype: float64

Checking mode for Chinstrap penguin weight

In [78]:
chinstrap_df.body_mass_g.mode()

0    3400.0
1    3650.0
2    3700.0
3    3800.0
4    3950.0
5    4050.0
Name: body_mass_g, dtype: float64

In [79]:
chinstrap_df.body_mass_g.median()

3700.0

Now looking at the mean, mode and median of the data split out for chinstrap penguins for example - we can see that they are almost equal so we could say the data is normally distributed. We would find similar results for both Gentoo and Adelie species if we split them out.

With this in mind, to model the data it would be best to split out by penguin species and then use the normal distribution as follows:   

mu, sigma = 0, 0.1 # mean and standard deviation   
s = np.random.default_rng().normal(mu, sigma, 1000)

In [80]:
# Model body_mass_g variable using normal distributions
rng = np.random.default_rng()

mu, sigma = 3700, 384.335081  # mean, standard deviation for Chinstrap penguins
body_mass_prob = np.random.default_rng().normal(mu, sigma, 1000)

# Print results if required
# print (f'Randam weights for Chinstrap penguins: \n{body_mass_prob}')

### Flipper Length

Flipper Length is also a continuous variable and would potentially also be normally distributed. Similar to body mass above we will follow the same process and check the variable firslty as a whole.

We can see that the mean and 50% percentile are relatively close but there is still significant deviation between the mean and the max values indicating skewness.

In [81]:
df.flipper_length_mm.describe()

count    342.000000
mean     200.915205
std       14.061714
min      172.000000
25%      190.000000
50%      197.000000
75%      213.000000
max      231.000000
Name: flipper_length_mm, dtype: float64

Taking just the Adelie species as an example...

In [82]:
fl_adelie_df = df.loc[df.loc[:,'species'] == 'Adelie']
fl_adelie_df.flipper_length_mm.describe()

count    151.000000
mean     189.953642
std        6.539457
min      172.000000
25%      186.000000
50%      190.000000
75%      195.000000
max      210.000000
Name: flipper_length_mm, dtype: float64

In [83]:
fl_adelie_df.flipper_length_mm.mode()

0    190.0
Name: flipper_length_mm, dtype: float64

In [84]:
fl_adelie_df.flipper_length_mm.median()

190.0

Now looking at the mean, mode and median of the data split out for Adelie penguins for example - we can see that they are exactly equal so the data is normally distributed. We would find similar results for both Gentoo and Adelie species if we split them out by flipper length.

With this in mind, to model the data it would be best to split out by penguin species and then use the normal distribution as follows:   

mu, sigma = 0, 0.1 # mean and standard deviation   
s = np.random.default_rng().normal(mu, sigma, 1000)

In [85]:
# Model flipper_length_mm variable using normal distributions
rng = np.random.default_rng()

mu, sigma = 190, 6.539457  # mean, standard deviation for Chinstrap penguins
flipper_length_prob = np.random.default_rng().normal(mu, sigma, 1000)

# Print results if required
# print (f'Randam flipper lengths for Adelie penguins: \n{flipper_length_prob}')

### Bill Length

Flipper Length is also a continuous variable and would potentially also be normally distributed. Similar to body mass above we will follow the same process and check the variable firslty as a whole...

...We can see that the mean and 50% percentile are relatively close but there is still significant deviation between the mean and the max values indicating skewness.

### Bill Depth

Bill Depth is also a continuous variable and would potentially also be normally distributed. Similar to body mass above we will follow the same process and check the variable firslty as a whole...

...We can see that the mean and 50% percentile are relatively close but there is still significant deviation between the mean and the max values indicating skewness.

# Task4

>The purpose of this task is to plot the entropy of the total number of heads versus the probability $p$ of giving heads when flipping two coins.

Firstly we get the probability of both coins returning a H for 10000 tosses and plot this for reference.

Note: we are making the assumption that the question refers to the probability of both coins landing on Heads with each single toss - i.e. 1 in 4 chance (**HH**, HT, TH, TT) as opposed to referring to either coin landing on a Head which would be 3 in 4 chance (**HH**, **HT**, **TH**, TT).

In [86]:
# Flipping two coins 10000 times to show the resulting probability of getting HH
rand_coins = np.random.binomial(2, .5, 10000)

# Plot the results
fig, ax = plt.subplots(figsize=(12, 4))
x, counts = np.unique(rand_coins, return_counts=True)
ax.bar(x, counts);

NameError: name 'plt' is not defined

In [ ]:
# Display the counts value

counts

As expected, the probability of HH comes in at approx .25 which we can verify with... 

In [ ]:
# probability would be .5 for each coin so combined would be
p = .50 * .50

p

In [ ]:
# Define the Entropy Formula as a function

def H(p):
    return -(1 - p) * np.log2(1.0 - p) - p * np.log2(p)

In [ ]:
# Calculate the Entropy value for the probability achieved above (i.e 25%)

H(0.25)

In [ ]:
# Plotting Entropy

fig, ax = plt.subplots(figsize= (8,4))

# Plot full range of probabilities from 0 to 1 and the associated Enttopy value
p = np.linspace(0.000001, 0.999999, 10001)
ax.plot(p, H(p));

# Add lines to indicate where our Probability (25%) and Entropy (0.81) meet
p=0.25 #Probability
e=0.81 #Entropy

plt.plot((p, p), (0.0, e),color='r');
plt.plot((0.0, p), (e, e),color='r');
plt.scatter(0.25, 0.81,color='r');
plt.title("Entropy v Probability for 2xH for two coin tosses")
plt.xlabel("Probability");
plt.ylabel("Entropy ");

# Task5

>The purpose of this task is to create approprite individual plots for each of the variables in the penguin data set.

The following are the variables in the dataset:

| Variable Name |
| --- |
| species |
| island |
| bill_length_mm |
| bill_depth_mm |
| flipper_length_mm |
| body_mass_g |
| sex |

Describe the data and give reason for the chosen plot type...
discrete v continuous, 
Notes 
Species - Pie/Bar Chart   
Island - Pie/Bar Chart   
Sex - Pie Chart   
flipper/bill size and mass - historgrams   

Scatter Plots for comparing x vs y etc



Sources:
https://www.kaggle.com/code/tirendazacademy/penguin-dataset-data-visualization-with-seaborn



### species
Species is a categorical variable and as we saw earlier, contains just three possible values - Adelie, Chinstrap and Gentoo. The most appropriate plot to use would be a bar chart showing the frequency/count of each species.

In [ ]:
# species

#Create a list just for Species and generate a bar chart

import matplotlib.pyplot as plt

species_l = []
for x in df['species']:
    species_l.append(x)

plt.title("Species")
plt.ylabel("Frequency")
temp=np.unique(species_l)
freq_list = []
for x in temp:
    freq_list.append(species_l.count(x))

plt.bar(temp, freq_list)
plt.show;

### island
Species is also a categorical variable and as we saw earlier, also contains just three possible values - Biscoe, Dream and Torgersen. The most appropriate plot to use would also be a bar chart showing the frequency/count of each island.

In [ ]:
# island

# Create a list just for island and generate a bar chart

import matplotlib.pyplot as plt

island_l = []
for x in df['island']:
    island_l.append(x)

plt.title("Island")
plt.ylabel("Frequency")
temp=np.unique(island_l)
freq_list = []
for x in temp:
    freq_list.append(island_l.count(x))

plt.bar(temp, freq_list)
plt.show;

### sex
Sex is another categorical variable and contains just two possible values - Male and Female. The most appropriate plot to use could be a bar chart showing the frequency/count of each sex or we could also use a pie chart showing the breakdown to include the percentages of male/female.

In [ ]:
# sex

# Create a list just for sex and generate a bar chart

import matplotlib.pyplot as plt

sex_l = []

for x in df['sex']:
    if str(x) != "nan":
        sex_l.append(x)

temp=np.unique(sex_l)
freq_list = []
for x in temp:
    freq_list.append(sex_l.count(x))

plt.title("Sex")
plt.pie(freq_list, labels = temp, autopct='%.2f')
plt.show;

### flipper_length
Flipper_length is a quantative variable with numeric values which are ordinal. The most interesting features for a set of data like this could be basic statistics such as min, max, mean and also the overall summary of data. The most appropriate plot to show these values would be a histogram. Given the variance by species it would make sense to plot the three species seperately rather than as a single data set.

In [ ]:
# Create a list just for flipper_length and create a histogram

flipper_length_gentoo_l=df.loc[df["species"]=="Gentoo"]
flipper_length_chinstrap_l=df.loc[df["species"]=="Chinstrap"]
flipper_length_adelie_l=df.loc[df["species"]=="Adelie"]

# Add title and labels
plt.title("Flipper Length")
plt.xlabel("Length (mm)")
plt.ylabel("Frequency")

# Create the plots
plt.hist(flipper_length_gentoo_l["flipper_length_mm"],label='Gentoo')
plt.hist(flipper_length_adelie_l["flipper_length_mm"],label='Adelie')
plt.hist(flipper_length_chinstrap_l["flipper_length_mm"],label='Chinstrap')
plt.legend();

### body_mass
Body_mass is also a quantative variable with numeric values which are ordinal. Again, the most interesting features for a set of data like this could be basic statistics such as min, max, mean and also the overall summary of data. The most appropriate plot to show these values would be a histogram. And, given the variance by species it would make sense to plot the three species seperately rather than as a single data set.

In [ ]:
# Create a list just for body_mass and create a histogram

body_mass_gentoo_l=df.loc[df["species"]=="Gentoo"]
body_mass_chinstrap_l=df.loc[df["species"]=="Chinstrap"]
body_mass_adelie_l=df.loc[df["species"]=="Adelie"]

# Add title and labels
plt.title("Body Mass")
plt.xlabel("Weight (g)")
plt.ylabel("Frequency")

# Create the plots
plt.hist(body_mass_gentoo_l["body_mass_g"],label='Gentoo')
plt.hist(body_mass_adelie_l["body_mass_g"],label='Adelie')
plt.hist(body_mass_chinstrap_l["body_mass_g"],label='Chinstrap')
plt.legend();

### bill_length
Bill_length is a quantative variable with numeric values which are ordinal. The most interesting features for a set of data like this could be basic statistics such as min, max, mean and also the overall summary of data. The most appropriate plot to show these values would be a histogram. Given the variance by species it would make sense to plot the three species seperately rather than as a single data set.

In [ ]:
# Create a list just for bill_length and create a histogram

bil_length_gentoo_l=df.loc[df["species"]=="Gentoo"]
bill_length_chinstrap_l=df.loc[df["species"]=="Chinstrap"]
bill_length_adelie_l=df.loc[df["species"]=="Adelie"]

# Add title and labels
plt.title("Bill Length")
plt.xlabel("Length (mm)")
plt.ylabel("Frequency")

# create the plots
plt.hist(bil_length_gentoo_l["bill_length_mm"],label='Gentoo')
plt.hist(bill_length_adelie_l["bill_length_mm"],label='Adelie')
plt.hist(bill_length_chinstrap_l["bill_length_mm"],label='Chinstrap')
plt.legend();


### bill depth
Bill_depth is likewise a quantative variable with numeric values which are ordinal. Again, the most interesting features for a set of data like this could be basic statistics such as min, max, mean and also the overall summary of data. The most appropriate plot to show these values would be a histogram and as above with bill_length, iven the variance by species it would make sense to plot the three species seperately rather than as a single data set.

In [ ]:
# Create a list just for bill_depth and create a histogram

bill_depth_gentoo_l=df.loc[df["species"]=="Gentoo"]
bill_depth_chinstrap_l=df.loc[df["species"]=="Chinstrap"]
bill_depth_adelie_l=df.loc[df["species"]=="Adelie"]

# Add title and labels
plt.title("Bill Depth")
plt.xlabel("Depth (mm)")
plt.ylabel("Frequency")

# Create the plots
plt.hist(bill_depth_gentoo_l["bill_depth_mm"],label='Gentoo')
plt.hist(bill_depth_adelie_l["bill_depth_mm"],label='Adelie')
plt.hist(bill_depth_chinstrap_l["bill_depth_mm"],label='Chinstrap')
plt.legend();

Given that bill_length and bill_depth are related, rather than showing them as individual plots as above, we could visualise the data using a scatterplot to give a true reflection of the relationship between the variables.

In [ ]:
# Create a scatter plot to compare bill_length and bill_depth

plt.scatter(df['bill_length_mm'], df['bill_depth_mm'], label='Bill Length | Bill Depth\n')
plt.title('Bill Length | Bill Depth')
plt.xlabel('Bill Length (mm)')
plt.ylabel('Bill Depth (mm)');

# References

$^1$ mwaskom/seaborn-data: Data repository for seaborn examples. Aug. 30, 2023. url: https://github.com/mwaskom/ seaborn - data / blob / master / penguins.csv (visited on 10/25/2023)


***

## End